### ISSUE TITLE
Implement language server protocol

### ISSUE THREAD

**Issue by andrewrk** [zig-lang/zig#615]
https://langserver.org

**Comment by binary132** [comment]
Does it make sense to tie this to https://github.com/zig-lang/zig/issues/89 or do you think separate pieces of the functionality required for LSP should be delegated into a pipeline of components which perhaps the core compiler functionality would also rely on?

**Comment by andrewrk** [comment]
My plan is to solve this only in the self-hosted compiler.

**Comment by bb010g** [comment]
I'd be interested in investigating the differences between the LSP and what the Dart [analyzer](https://github.com/dart-lang/sdk/tree/master/pkg/analyzer_cli) ([API spec](https://htmlpreview.github.io/?https://github.com/dart-lang/sdk/blob/master/pkg/analysis_server/doc/api.html)) provides, as JetBrains is actually able to use Dart's analyzer to fully power their [Dart support](https://www.jetbrains.com/help/idea/dart.html) due to the [different features](https://www.reddit.com/r/rust/comments/5rhqj0/ide_compiler_has_very_different_requirements_and/dd8khei/) provided. Additionally, it looks like you can [shim](https://github.com/natebosch/dart_language_server) Dart's Analysis Server API to provide the Language Server Protocol without too much trouble.

**Comment by andrewrk** [comment]
I know @thejoshwolfe has some bones to pick with the language server protocol in particular. I think in the end we will likely support a custom protocol that is more powerful than LSP, and theoretically you could implement a proxy that consumes the former and provides the latter. Perhaps zig can support both.

**Comment by monouser7dig** [comment]
> I know @thejoshwolfe has some bones to pick with the language server protocol in particular.

No matter what, this will make working with zig so much easier 🥇 
1. editor completion support but also the whole ecosystem 

one more example:
really nice code navigation https://about.sourcegraph.com/blog/sourcegraph-is-now-open-source/

**Comment by andrewrk** [comment]
This isn't contributor friendly yet. Lots to solve before work on this begins.

**Comment by RUSshy** [comment]
I found this https://github.com/ice1000/intellij-zig
It looks promising

**Comment by aindigo** [comment]
@andrewk is this your idea?:   to have a language server implemented as a function in the std lib, so can be called from the build script. and then spawn the process with 'zig run-lsp ...args'. I think some basic methods can be implemented from the LSP spec with what zig have right now.

**Comment by andrewk** [comment]
You pinged the wrong account.

**Comment by ice1000** [comment]
> I found this https://github.com/ice1000/intellij-zig
> It looks promising

Maybe not, hasn't been updated for a while

**Comment by RUSshy** [comment]
why?

**Comment by ice1000** [comment]
Because I'm busy

**Comment by ice1000** [comment]
I haven't implemented complicated reference resolution, thus you cannot have cross-file jumps and type inference. Zig is really unfriendly for IDE due to the C-interop hacks (you can use C references, which requires your IDE to understand C. For a language server, you also need to provide C source information, while you have LLVM in the Zig compiler to make things easier).

**Comment by xackus** [comment]
Please be aware that for best results the language parser should be error tolerant.
https://code.visualstudio.com/api/language-extensions/language-server-extension-guide#error-tolerant-parser-for-language-server

Currently the JSON dumps @andrewrk suggested to use in #3429 are not generated in case of a compilation error.
I do not know the compiler internals, how feasible is error tolerance when generating the JSON dump in stage1 and stage2 compiler?

Also, I presume the JSON format isn't planned to drastically change in the self-hosted compiler, is that correct?
I would be happy to start writing the language server.
In what directory in the repo should the server code be placed?

**Comment by ice1000** [comment]
You can don't have an error-tolerant parser.
You can do: obtain AST each time when the code is well-formed, listen to programmer typing dots or explicitly invoke completion, and use the AST you obtained just now to provide completion.

Although this is definitely less good than error-tolerant parser (which is available in the unmaintained intellij-zig).

**Comment by xackus** [comment]
Correct. I just wanted make sure that new code is written to allow for error tolerance if possible.

**Comment by xackus** [comment]
I am currently implementing JSON-RPC.

**Comment by SuperAuguste** [comment]
If anyone is still actively looking for a language server, I'm one of the maintainers of `zls`, an almost full featured language server for Zig! It has local, global, and package completions and tons of other features such as goto, hover definitions, and document outlines coming soon.

You can check it out [here](https://github.com/zigtools/zls).

**Comment by matklad** [comment]
>This will not be [LSP](https://langserver.org/), but a custom protocol just for Zig. A third party project can act as a proxy between this and LSP for that use case.

Two thoughts:

1) consider building a custom library (with well-isolated pure-data shaped API) rather than an IPC protocol. LSP is all the rage nowadays, but better not to close the avenue for embedding the thing in process. 
2) +1 on not coding against LSP specifically. The design principle we use for API layer in rust-analyzer is a) being 100% rust-specific b) targeting hypothetical “perfect” client which takes full advantage of what we provide. That general API is then explicitly dumbed down to LSP’s capabilities.

**Comment by Jarred-Sumner** [comment]
The current state as of January 2023 when you try to autocomplete `sockaddr_t` (from #14209)

#### Zig

1. Type `std.os.sockaddr`
2. No completions appear. Data types don't either. Go-to definition doesn't help

https://user-images.githubusercontent.com/709451/210762562-1d4befe6-ea46-4547-88bb-d8e9bf577d50.mp4

#### Rust

When you use Rust, completions work relatively consistently. In-editor feedback reports when the wrong number of arguments are passed to a function. It doesn't take more than a second or two to update the red squiggly lines. This improves developer productivity.

https://user-images.githubusercontent.com/709451/210764636-91f4c140-822b-4229-9d67-32f5fa6fb385.mp4

#### Go

It works pretty well, including go-to definition

https://user-images.githubusercontent.com/709451/210766932-f4d3432e-890c-4c89-a3bf-80350c04b454.mp4

**Comment by SuperAuguste** [comment]
@Jarred-Sumner Keep in mind that this is the current state of completions in zls and not the current state of completions in Zig; the compiler server is not implemented yet.

**Comment by Jarred-Sumner** [comment]
> @Jarred-Sumner Keep in mind that this is the current state of completions in zls and not the current state of completions in Zig; the compiler server is not implemented yet.

Sure, but it is the current state of autocomplete for developers using Zig. It's relevant for this thread because it is an infrastructure/tooling problem that seems difficult for zls to solve well alone. [`rust-analyzer`](https://github.com/rust-lang/rust-analyzer) describes itself as a "Rust compiler frontend for IDEs" and that is essentially what this thread proposes (except, for Zig).

**Comment by matklad** [comment]
Some high-level thoughts on how one could actually do this:

https://matklad.github.io/2023/02/10/how-a-zig-ide-could-work.html

**Comment by andrewrk** [comment]
Thank you for sharing! This dives a bit deeper than my own musings on the topic, so I found it quite insightful. 

> If you think about IDE support for the first time, a very tantalizing idea is to try to lean just into incremental compilation.

Guilty as charged 😅 and excellent points thereafter. Although I do think that may be a nice starting point - especially with some of the upcoming compilation speed enhancements that I have in mind.
